File: snippets/security/authentication/oauth2_authorization_code.py
"""
OAuth2 Authorization Code Flow Implementation
Production-ready OAuth2 with PKCE support
"""
import secrets
import hashlib
import base64
from typing import Dict, Optional, Tuple
from urllib.parse import urlencode, parse_qs, urlparse
import requests


class OAuth2Client:
    """OAuth2 client with Authorization Code Flow and PKCE"""

    def __init__(
        self,
        client_id: str,
        client_secret: str,
        authorization_url: str,
        token_url: str,
        redirect_uri: str,
        scope: Optional[str] = None
    ):
        self.client_id = client_id
        self.client_secret = client_secret
        self.authorization_url = authorization_url
        self.token_url = token_url
        self.redirect_uri = redirect_uri
        self.scope = scope or "openid profile email"

    def generate_pkce_pair(self) -> Tuple[str, str]:
        """
        Generate PKCE code verifier and challenge

        Returns:
            Tuple of (code_verifier, code_challenge)
        """
        # Generate code verifier (43-128 characters)
        code_verifier = base64.urlsafe_b64encode(
            secrets.token_bytes(32)
        ).decode('utf-8').rstrip('=')

        # Generate code challenge (SHA256 hash of verifier)
        code_challenge = base64.urlsafe_b64encode(
            hashlib.sha256(code_verifier.encode('utf-8')).digest()
        ).decode('utf-8').rstrip('=')

        return code_verifier, code_challenge

    def get_authorization_url(
        self,
        state: Optional[str] = None,
        use_pkce: bool = True
    ) -> Tuple[str, Dict[str, str]]:
        """
        Generate authorization URL

        Args:
            state: Optional state parameter for CSRF protection
            use_pkce: Whether to use PKCE

        Returns:
            Tuple of (authorization_url, session_data)
        """
        state = state or secrets.token_urlsafe(32)
        session_data = {"state": state}

        params = {
            "client_id": self.client_id,
            "redirect_uri": self.redirect_uri,
            "response_type": "code",
            "scope": self.scope,
            "state": state
        }

        if use_pkce:
            code_verifier, code_challenge = self.generate_pkce_pair()
            params["code_challenge"] = code_challenge
            params["code_challenge_method"] = "S256"
            session_data["code_verifier"] = code_verifier

        auth_url = f"{self.authorization_url}?{urlencode(params)}"
        return auth_url, session_data

    def exchange_code_for_token(
        self,
        code: str,
        code_verifier: Optional[str] = None
    ) -> Optional[Dict]:
        """
        Exchange authorization code for access token

        Args:
            code: Authorization code from callback
            code_verifier: PKCE code verifier if used

        Returns:
            Token response dictionary
        """
        data = {
            "grant_type": "authorization_code",
            "code": code,
            "redirect_uri": self.redirect_uri,
            "client_id": self.client_id,
            "client_secret": self.client_secret
        }

        if code_verifier:
            data["code_verifier"] = code_verifier

        try:
            response = requests.post(
                self.token_url,
                data=data,
                headers={"Content-Type": "application/x-www-form-urlencoded"},
                timeout=10
            )
            response.raise_for_status()
            return response.json()
        except requests.exceptions.RequestException as e:
            print(f"Token exchange failed: {e}")
            return None

    def refresh_access_token(self, refresh_token: str) -> Optional[Dict]:
        """
        Refresh access token using refresh token

        Args:
            refresh_token: Refresh token

        Returns:
            New token response
        """
        data = {
            "grant_type": "refresh_token",
            "refresh_token": refresh_token,
            "client_id": self.client_id,
            "client_secret": self.client_secret
        }

        try:
            response = requests.post(
                self.token_url,
                data=data,
                timeout=10
            )
            response.raise_for_status()
            return response.json()
        except requests.exceptions.RequestException as e:
            print(f"Token refresh failed: {e}")
            return None

    def revoke_token(self, token: str, token_type_hint: str = "access_token") -> bool:
        """
        Revoke access or refresh token

        Args:
            token: Token to revoke
            token_type_hint: Type of token (access_token or refresh_token)

        Returns:
            True if successful
        """
        revocation_url = self.token_url.replace("/token", "/revoke")

        data = {
            "token": token,
            "token_type_hint": token_type_hint,
            "client_id": self.client_id,
            "client_secret": self.client_secret
        }

        try:
            response = requests.post(revocation_url, data=data, timeout=10)
            return response.status_code == 200
        except requests.exceptions.RequestException:
            return False


# Example usage
if __name__ == "__main__":
    oauth_client = OAuth2Client(
        client_id="your_client_id",
        client_secret="your_client_secret",
        authorization_url="https://provider.com/oauth/authorize",
        token_url="https://provider.com/oauth/token",
        redirect_uri="https://yourapp.com/callback"
    )

    # Step 1: Get authorization URL
    auth_url, session = oauth_client.get_authorization_url(use_pkce=True)
    print(f"Authorization URL: {auth_url}")
    print(f"Session data: {session}")

    # Step 2: User authorizes and you receive callback with code
    # code = "authorization_code_from_callback"
    # tokens = oauth_client.exchange_code_for_token(
    #     code,
    #     code_verifier=session.get("code_verifier")
    # )

File: snippets/security/authentication/oauth2_authorization_code.ts
/**
 * OAuth2 Authorization Code Flow Implementation (TypeScript)
 * Production-ready OAuth2 with PKCE support
 */
import { randomBytes, createHash } from 'crypto';
import axios, { AxiosError } from 'axios';

interface OAuth2Config {
  clientId: string;
  clientSecret: string;
  authorizationUrl: string;
  tokenUrl: string;
  redirectUri: string;
  scope?: string;
}

interface PKCEPair {
  codeVerifier: string;
  codeChallenge: string;
}

interface SessionData {
  state: string;
  codeVerifier?: string;
}

interface TokenResponse {
  access_token: string;
  token_type: string;
  expires_in: number;
  refresh_token?: string;
  scope?: string;
}

export class OAuth2Client {
  private config: Required<OAuth2Config>;

  constructor(config: OAuth2Config) {
    this.config = {
      ...config,
      scope: config.scope || 'openid profile email',
    };
  }

  /**
   * Generate PKCE code verifier and challenge
   * @returns Object with code verifier and challenge
   */
  generatePKCEPair(): PKCEPair {
    // Generate code verifier (43-128 characters)
    const codeVerifier = randomBytes(32)
      .toString('base64url');

    // Generate code challenge (SHA256 hash of verifier)
    const codeChallenge = createHash('sha256')
      .update(codeVerifier)
      .digest('base64url');

    return { codeVerifier, codeChallenge };
  }

  /**
   * Generate authorization URL
   * @param state - Optional state for CSRF protection
   * @param usePKCE - Whether to use PKCE
   * @returns Authorization URL and session data
   */
  getAuthorizationUrl(
    state?: string,
    usePKCE: boolean = true
  ): { url: string; sessionData: SessionData } {
    const stateValue = state || randomBytes(32).toString('base64url');
    const sessionData: SessionData = { state: stateValue };

    const params = new URLSearchParams({
      client_id: this.config.clientId,
      redirect_uri: this.config.redirectUri,
      response_type: 'code',
      scope: this.config.scope,
      state: stateValue,
    });

    if (usePKCE) {
      const { codeVerifier, codeChallenge } = this.generatePKCEPair();
      params.append('code_challenge', codeChallenge);
      params.append('code_challenge_method', 'S256');
      sessionData.codeVerifier = codeVerifier;
    }

    const url = `${this.config.authorizationUrl}?${params.toString()}`;
    return { url, sessionData };
  }

  /**
   * Exchange authorization code for access token
   * @param code - Authorization code from callback
   * @param codeVerifier - PKCE code verifier if used
   * @returns Token response
   */
  async exchangeCodeForToken(
    code: string,
    codeVerifier?: string
  ): Promise<TokenResponse | null> {
    const data = new URLSearchParams({
      grant_type: 'authorization_code',
      code,
      redirect_uri: this.config.redirectUri,
      client_id: this.config.clientId,
      client_secret: this.config.clientSecret,
    });

    if (codeVerifier) {
      data.append('code_verifier', codeVerifier);
    }

    try {
      const response = await axios.post<TokenResponse>(
        this.config.tokenUrl,
        data.toString(),
        {
          headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
          timeout: 10000,
        }
      );

      return response.data;
    } catch (error) {
      if (error instanceof AxiosError) {
        console.error('Token exchange failed:', error.message);
      }
      return null;
    }
  }

  /**
   * Refresh access token using refresh token
   * @param refreshToken - Refresh token
   * @returns New token response
   */
  async refreshAccessToken(
    refreshToken: string
  ): Promise<TokenResponse | null> {
    const data = new URLSearchParams({
      grant_type: 'refresh_token',
      refresh_token: refreshToken,
      client_id: this.config.clientId,
      client_secret: this.config.clientSecret,
    });

    try {
      const response = await axios.post<TokenResponse>(
        this.config.tokenUrl,
        data.toString(),
        { timeout: 10000 }
      );

      return response.data;
    } catch (error) {
      if (error instanceof AxiosError) {
        console.error('Token refresh failed:', error.message);
      }
      return null;
    }
  }

  /**
   * Revoke access or refresh token
   * @param token - Token to revoke
   * @param tokenTypeHint - Type of token
   * @returns True if successful
   */
  async revokeToken(
    token: string,
    tokenTypeHint: 'access_token' | 'refresh_token' = 'access_token'
  ): Promise<boolean> {
    const revocationUrl = this.config.tokenUrl.replace('/token', '/revoke');

    const data = new URLSearchParams({
      token,
      token_type_hint: tokenTypeHint,
      client_id: this.config.clientId,
      client_secret: this.config.clientSecret,
    });

    try {
      await axios.post(revocationUrl, data.toString(), { timeout: 10000 });
      return true;
    } catch {
      return false;
    }
  }
}

// Example usage
if (require.main === module) {
  const oauthClient = new OAuth2Client({
    clientId: 'your_client_id',
    clientSecret: 'your_client_secret',
    authorizationUrl: 'https://provider.com/oauth/authorize',
    tokenUrl: 'https://provider.com/oauth/token',
    redirectUri: 'https://yourapp.com/callback',
  });

  // Step 1: Get authorization URL
  const { url, sessionData } = oauthClient.getAuthorizationUrl(undefined, true);
  console.log('Authorization URL:', url);
  console.log('Session data:', sessionData);
}
